### PR TITLE
CRM_Contact_Form_Contact to use addField CRM-16178

### PR DESCRIPTION
**Pull request by mallezie**
This converts CRM_Contact_Form_Contact to use addField. This cleans up the select element, and adds support for file elements.

### CODE REVIEW

File: CRM/Contact/Form/Contact.php [hunk 1]
```diff
@@ -818,8 +818,7 @@ public function buildQuickForm() {
     CRM_Contact_Form_Location::buildQuickForm($this);
 
     // add attachment
-    $this->addElement('file', 'image_URL', ts('Browse/Upload Image'), 'size=30 maxlength=60');
-    $this->addUploadElement('image_URL');
+    $this->addField('image_URL', array('type' => 'File', 'size' => '30', 'maxlength' => '60', 'label' => ts('Browse/Upload Image')));
```
Comment (colemanw) [inline CRM/Contact/Form/Contact.php]: `'type' => 'File'` can probably be defined in the schema, as well as size and maxlength.

File: CRM/Core/Form.php [hunk 2]
```diff
@@ -1249,13 +1254,22 @@ public function addField($name, $props = array(), $required = FALSE) {
         if (empty($props['multiple'])) {
           $options = array('' => $props['placeholder']) + $options;
         }
-        $this->add('select', $name, $label, $options, $required, $props);
+        $props += array('required' => $required);
+        $this->addElement('select', $name, $label, $options, $props);
```
Comment (colemanw) [inline CRM/Core/Form.php]: I don't understand this change. The `add` method enforces the required rule. AFAIK `addElement` does not.

File: CRM/Core/Form.php [hunk 3]
```diff
@@ -1249,13 +1254,22 @@ public function addField($name, $props = array(), $required = FALSE) {
...
 
       //case 'AdvMulti-Select':
       //case 'CheckBox':
-      //case 'File':
+      case 'File':
+        // We should not build upload file in search mode.
+        if (isset($props['context']) && $props['context'] == 'search') {
+          return;
+        }
+        $this->addElement('file', $name, $label, $props, $required);
```
Comment (colemanw) [inline CRM/Core/Form.php]: I think we need to switch `addElement` to `add` so the required rule will be properly set.

**Comment by mallezie** [comment]
I missed the required rules in the add-method i see now. I was actually trying to avoid the add-method, and try to make clean use of addElement. There are lot's of add methods in the Form class i would like to clean up after this change. I somehow see the addField method to become something like this. 
```
in addField(){
// Define properties and widget as now. 
// Replace add-methods and the case to something like this:
$this->add . '$widget' . method( with properties)
}

// Create add widget methods most are already there (could be even private, but i wouldn't mark them that way however). For example
$this-addText(){
  $this-AddElement();
  $this->AddRule(required)
 // Add other rules defined on schema
}
```
In the end we could then have addField for all methods, addElement if we want to add a custom form element, and some helper methods like addButtons and stuff. And then we could remove some addfunctions, or at least change the forms to only use addField / addElement or addHelper (like addButton). Which would make all the form classes easier. Not sure however this is a feasible approach. 

In this patch this would mean i do need to add the required rules. I'll do that first, then we can discuss the approach.

**Comment by mallezie** [comment]
I hope the last commit somehow shows how i would see it.

**Comment by mallezie** [comment]
I sort of don't really like the add-method since it does multiple thing. It adds the required rule, and multiple for selects, and the placeholder for selects, and it does something else with errors to the form. That's too much, so i would like to split that in the long term.

**Comment by mallezie** [comment]
Any feedback on this proposed approach?

**Comment by colemanw** [comment]
Hmm, we might just have to agree to disagree about CRM_Core_Form::add(). IMO the 3 things it does are all useful and necessary, and I don't see the case for duplicating the functionality elsewhere.

To clarify, is there something it's doing that you *don't* want it to do and therefore it's getting in your way?

If not, I'd vote for using it from addField and save further refactoring for down the road.

**Comment by mallezie** [comment]
@colemanw thanks for the reply. I can see it's better to save the ```add``` function refactoring for later. 
It's nothing which is getting in my way, at first it was just a bit confusing to see the difference between ```add``` and ```addElement```. The only 'problem' i have with the 'add' method is that it does 3 'different' things, but that's perhaps something for a followup. My aim was not to duplicate the functionality but split it up in it's 3 different components. 

I'll revert the add and addElement changes over the weekend.

**Comment by mallezie** [comment]
I reverted the addElement changes.

**Pull request merged**